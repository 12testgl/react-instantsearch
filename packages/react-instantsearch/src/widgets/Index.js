import createIndex from '../core/createIndex';

const Index = createIndex({
  Root: 'div',
  props: { className: 'ais-MultiIndex__root' },
});
export default Index;
