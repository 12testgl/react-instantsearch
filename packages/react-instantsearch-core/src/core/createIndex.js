import React from 'react';
import PropTypes from 'prop-types';
import Index from '../components/Index';

/**
 * Creates a specialized root Index component. It accepts
 * a specification of the root Element.
 * @param {object} defaultRoot - the defininition of the root of an Index sub tree.
 * @return {object} a Index root
 */
const createIndex = defaultRoot => {
  let hasAlreadyWarn = false;
  const CreateIndex = ({ indexName, indexId, root, children }) => {
    if (process.env.NODE_ENV !== 'production' && !hasAlreadyWarn && !indexId) {
      hasAlreadyWarn = true;
      // eslint-disable-next-line no-console
      console.warn(
        '[React InstantSearch]: `indexId` is required for the `Index` component. Please use this prop before the next major version.'
      );
    }

    return (
      <Index indexName={indexName} indexId={indexId || indexName} root={root}>
        {children}
      </Index>
    );
  };

  CreateIndex.propTypes = {
    indexName: PropTypes.string.isRequired,
    // @MAJOR: indexId must be required
    indexId: PropTypes.string,
    root: PropTypes.shape({
      Root: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.func,
        PropTypes.object,
      ]).isRequired,
      props: PropTypes.object,
    }),
    children: PropTypes.node,
  };

  CreateIndex.defaultProps = {
    root: defaultRoot,
  };

  return CreateIndex;
};

export default createIndex;
