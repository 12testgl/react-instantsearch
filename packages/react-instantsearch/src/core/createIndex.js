import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Index from './Index';

/**
 * Creates a specialized root Index component. It accepts
 * a specification of the root Element.
 * @param {object} root - the defininition of the root of an Index sub tree.
 * @returns {object} a Index root
 */
export default function createIndex(root) {
  return class CreateIndex extends Component {
    static propTypes = {
      indexName: PropTypes.string.isRequired,
      root: PropTypes.shape({
        Root: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
          .isRequired,
        props: PropTypes.object,
      }),
      children: PropTypes.node,
    };

    static defaultProps = {
      root,
    };

    render() {
      return (
        <Index indexName={this.props.indexName} root={this.props.root}>
          {this.props.children}
        </Index>
      );
    }
  };
}
