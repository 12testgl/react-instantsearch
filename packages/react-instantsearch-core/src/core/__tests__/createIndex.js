import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Index from '../../components/Index';
import createIndex from '../createIndex';

Enzyme.configure({ adapter: new Adapter() });

describe('createIndex', () => {
  const requiredProps = {
    indexName: 'indexName',
  };

  it('expects to create an Index', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.is(Index)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('expects to create an Index with the default root', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().root).toEqual({
      Root: 'div',
    });
  });

  it('expects to create an Index with a custom root props', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const props = {
      ...requiredProps,
      root: {
        Root: 'span',
        props: {
          style: {
            flex: 1,
          },
        },
      },
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().root).toEqual({
      Root: 'span',
      props: {
        style: {
          flex: 1,
        },
      },
    });
  });

  it('expects to create an Index with an indexId when provided', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const props = {
      ...requiredProps,
      indexId: 'indexId',
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().indexId).toBe('indexId');
  });

  it("expects to create an Index with an indexId that fallback to indexName when it's not provided", () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().indexId).toBe('indexName');
  });

  it('expect to warn when an Index is used without an indexId', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const warn = jest.spyOn(console, 'warn');

    const props = {
      ...requiredProps,
    };

    shallow(<CustomIndex {...props} />);

    expect(warn).toHaveBeenCalledWith(
      '[React InstantSearch]: `indexId` is required for the `Index` component. Please use this prop before the next major version.'
    );

    warn.mockRestore();
  });

  it('expect to warn only once when an Index is used without an indexId', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const warn = jest.spyOn(console, 'warn');

    const props = {
      ...requiredProps,
    };

    shallow(<CustomIndex {...props} />);
    shallow(<CustomIndex {...props} />);
    shallow(<CustomIndex {...props} />);

    expect(warn).toHaveBeenCalledTimes(1);

    warn.mockRestore();
  });

  it('expect to not warn when an Index is used with an indexId', () => {
    const CustomIndex = createIndex({ Root: 'div' });

    const warn = jest.spyOn(console, 'warn');

    const props = {
      ...requiredProps,
      indexId: 'indexId',
    };

    shallow(<CustomIndex {...props} />);

    expect(warn).not.toHaveBeenCalled();

    warn.mockRestore();
  });
});
