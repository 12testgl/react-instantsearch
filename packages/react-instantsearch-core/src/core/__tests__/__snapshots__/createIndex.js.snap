// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIndex expect to create an Index 1`] = `
<Index
  indexId="indexName"
  indexName="indexName"
  root={
    Object {
      "Root": "div",
    }
  }
/>
`;
